### PR TITLE
Distributed Session: Persist only the accessed arguments in the Datastore

### LINKED ISSUE #96: Distributed sessions: Serialize attributes in separate entities
*Current state*
After a request all the attributes and metadata attached to the session are serialized and stored in one entity (per session).

*Proposal*
Stores the attributes in separate entities, only update the accessed attributes and metadata.  

*Details*
* The operations for each session can be [Batched](https://cloud.google.com/datastore/docs/concepts/entities#batch_operations) so only one call to the datastore will be made.
* All the attributes for one session will be considered as one [Entity Group](https://cloud.google.com/datastore/docs/concepts/entities#entity_groups)

### PR DESCRIPTION
**Pull request by cassand**
This module adds a proxy on top of the session which keep track of the accessed attributes, and only send to the Datastore the modified attributes. 

This module can be activated using the property `gcp.distributed-sessions.useUniqueEntity`. 

The activation of this module implies that the attributes of the session are no longer stored in the same entity as the session metadata. In the case where the module is not active the behavior is the same as before (the session and its attributes are stored in the same entity).

All the operations of this modules can be traced and send to Stackdriver, by using the property `gcp.distributed-sessions.enableTrace`.

Fixes #96

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -78,6 +78,8 @@ The distributed sessions module can be configured through the environment variab
 | gcp.distributed-sessions.namespace    |  Namespace to use in the Datastore.                         |  tomcat-gcp-persistent-session |
 | gcp.distributed-sessions.sessionKind  |  Name of the entity used to store sessions in the Datastore. |  TomcatGCloudSession |
 | gcp.distributed-sessions.uriExcludePattern | [Pattern](https://docs.oracle.com/javase/8/docs/api/java/util/regex/Pattern.html) specifying which URI to ignore when persisting sessions. | null |
+| gcp.distributed-sessions.enableTrace | Send traces from the Datastore operations to Stackdriver. | false |
```
Comment (meltsufin) [inline README.md]: All datastore operations or just for session persistence? Can we make the description more clear?

File: README.md [hunk 2]
```diff
@@ -78,6 +78,8 @@ The distributed sessions module can be configured through the environment variab
...
+| gcp.distributed-sessions.useUniqueEntity | Regroup the session metadata and attributes in the same entity in the Datastore | true |
```
Comment (meltsufin) [inline README.md]: I think we need a more detailed description of the purpose and implications of using this setting. What does setting it to `false` do?

File: tomcat-gcp-lib/src/main/java/com/google/cloud/runtimes/tomcat/session/DatastoreStore.java [hunk 3]
```diff
@@ -72,6 +80,17 @@
   private String namespace;
 
   /**
+   * If true the session attributes are in the same entity as the session metadata, otherwise,
+   * each attribute is saved in a distinct entity.
+   */
+  private boolean useUniqueEntity = true;
```
Comment (meltsufin) [inline tomcat-gcp-lib/src/main/java/com/google/cloud/runtimes/tomcat/session/DatastoreStore.java]: `useSingleEntity`? We need to find a better name for this member.

**Review by meltsufin — COMMENTED** [review]
First pass...

I think there is a lot of complexity added by supporting two ways of storing the sessions via the `useUniqueEntity` option. What are the pros and cons of each? Can we just pick one?

File: tomcat-gcp-lib/src/main/java/com/google/cloud/runtimes/tomcat/session/DatastoreManager.java [hunk 4]
```diff
@@ -91,10 +95,13 @@ protected synchronized void startInternal() throws LifecycleException {
   @Override
   public Session findSession(String id) throws IOException {
     log.debug("Datastore manager is loading session: " + id);
+    Tracer tracer = Trace.getTracer();
```
Comment (balopat) [inline tomcat-gcp-lib/src/main/java/com/google/cloud/runtimes/tomcat/session/DatastoreManager.java]: I thought tracing is optional, and user should be able to opt-out of StackDriver trace, if they want to. 
Isn't that the `gcp.distributed-sessions.enableTrace` is for?

**Review by balopat — CHANGES_REQUESTED** [review]
Got an error with `mvn clean verify -P gcp-integration-test`:

````
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running com.google.cloud.runtimes.tomcat.session.integration.DatastoreStoreIT
[ERROR] Tests run: 8, Failures: 3, Errors: 0, Skipped: 0, Time elapsed: 6.659 s <<< FAILURE! - in com.google.cloud.runtimes.tomcat.session.integration.DatastoreStoreIT
[ERROR] testSessionEnumeration(com.google.cloud.runtimes.tomcat.session.integration.DatastoreStoreIT)  Time elapsed: 0.66 s  <<< FAILURE!
java.lang.AssertionError
	at com.google.cloud.runtimes.tomcat.session.integration.DatastoreStoreIT.testSessionEnumeration(DatastoreStoreIT.java:136)

[ERROR] testStoreReset(com.google.cloud.runtimes.tomcat.session.integration.DatastoreStoreIT)  Time elapsed: 0.626 s  <<< FAILURE!
java.lang.AssertionError: expected null, but was:<Entity{key=Key{projectId=ja-workspace-0, namespace=tomcat-gcp-persistent-session-test, path=[PathElement{kind=TomcatGCloudSession, id=null, name=123}]}, properties={attributes=ListValue{valueType=LIST, excludeFromIndexes=false, meaning=0, value=[]}, creationTime=LongValue{valueType=LONG, excludeFromIndexes=false, meaning=0, value=1505836770706}, expiration_time=LongValue{valueType=LONG, excludeFromIndexes=false, meaning=0, value=1505838570706}, isNew=BooleanValue{valueType=BOOLEAN, excludeFromIndexes=false, meaning=0, value=true}, isValid=BooleanValue{valueType=BOOLEAN, excludeFromIndexes=false, meaning=0, value=true}, lastAccessedTime=LongValue{valueType=LONG, excludeFromIndexes=false, meaning=0, value=1505836770706}, maxInactiveInterval=LongValue{valueType=LONG, excludeFromIndexes=false, meaning=0, value=1800}, thisAccessedTime=LongValue{valueType=LONG, excludeFromIndexes=false, meaning=0, value=1505836770706}}}>
	at com.google.cloud.runtimes.tomcat.session.integration.DatastoreStoreIT.testStoreReset(DatastoreStoreIT.java:173)

[ERROR] testSessionCount(com.google.cloud.runtimes.tomcat.session.integration.DatastoreStoreIT)  Time elapsed: 1.245 s  <<< FAILURE!
java.lang.AssertionError: expected:<2> but was:<1>
	at com.google.cloud.runtimes.tomcat.session.integration.DatastoreStoreIT.testSessionCount(DatastoreStoreIT.java:128)

[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   DatastoreStoreIT.testSessionCount:128 expected:<2> but was:<1>
[ERROR]   DatastoreStoreIT.testSessionEnumeration:136
[ERROR]   DatastoreStoreIT.testStoreReset:173 expected null, but was:<Entity{key=Key{projectId=ja-workspace-0, namespace=tomcat-gcp-persistent-session-test, path=[PathElement{kind=TomcatGCloudSession, id=null, name=123}]}, properties={attributes=ListValue{valueType=LIST, excludeFromIndexes=false, meaning=0, value=[]}, creationTime=LongValue{valueType=LONG, excludeFromIndexes=false, meaning=0, value=1505836770706}, expiration_time=LongValue{valueType=LONG, excludeFromIndexes=false, meaning=0, value=1505838570706}, isNew=BooleanValue{valueType=BOOLEAN, excludeFromIndexes=false, meaning=0, value=true}, isValid=BooleanValue{valueType=BOOLEAN, excludeFromIndexes=false, meaning=0, value=true}, lastAccessedTime=LongValue{valueType=LONG, excludeFromIndexes=false, meaning=0, value=1505836770706}, maxInactiveInterval=LongValue{valueType=LONG, excludeFromIndexes=false, meaning=0, value=1800}, thisAccessedTime=LongValue{valueType=LONG, excludeFromIndexes=false, meaning=0, value=1505836770706}}}>
[INFO] 
[ERROR] Tests run: 8, Failures: 3, Errors: 0, Skipped: 0
[INFO] 
[INFO] 
[INFO] --- maven-failsafe-plugin:2.20:verify (default) @ tomcat-gcp-lib ---
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] Google Tomcat Runtime Parent ....................... SUCCESS [  1.033 s]
[INFO] tomcat-gcp-lib ..................................... FAILURE [ 21.084 s]
[INFO] Google Tomcat Runtime Docker Image ................. SKIPPED
[INFO] Tomcat-Runtime :: Tests :: Simple Tests ............ SKIPPED
[INFO] Tomcat-Runtime :: Tests :: Integration Tests ....... SKIPPED
[INFO] Tomcat-Runtime :: Tests :: Spring boot tests ....... SKIPPED
[INFO] Tomcat-Runtime :: Tests :: Distributed Tests ....... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 22.360 s
[INFO] Finished at: 2017-09-19T11:59:33-04:00
[INFO] Final Memory: 51M/290M
[INFO] ------------------------------------------------------------------------

````

File: tests/test-distributed-war/src/main/docker/Dockerfile [hunk 5]
```diff
@@ -17,5 +17,5 @@ ADD ${project.build.finalName}.war $APP_DESTINATION
 COPY credential.json /credentials/key.json
 ENV GOOGLE_APPLICATION_CREDENTIALS /credentials/key.json
 
-ENV TOMCAT_MODULES_ENABLE distributed-sessions
-ENV TOMCAT_LOGGING_PROPERTIES com.google.cloud.runtimes.tomcat.session.level=ALL
\ No newline at end of file
+ENV TOMCAT_MODULES_ENABLE distributed-sessions,stackdriver-trace
+ENV TOMCAT_LOGGING_PROPERTIES com.google.cloud.runtimes.tomcat.session.level=ALL,gcp.distributed-sessions.enableTrace=true
```
Comment (balopat) [inline tests/test-distributed-war/src/main/docker/Dockerfile:21]: there is no integration test for `gcp.distributed-sessions.separateAttributes` - I'd add another instance with that turned on as well.

File: tomcat-gcp-lib/src/main/java/com/google/cloud/runtimes/tomcat/session/DatastoreManager.java (hunk 4 above)
Comment (cassand) [reply tomcat-gcp-lib/src/main/java/com/google/cloud/runtimes/tomcat/session/DatastoreManager.java]: By default a trace service which ignore all the traces is instancied (`NoTraceService` in the [Trace implementation](https://github.com/GoogleCloudPlatform/cloud-trace-java/blob/master/sdk/service/src/main/java/com/google/cloud/trace/Trace.java)).
Enabling the module `stackdriver-trace` replace this service with a GrpcTraceService, which send the traces to Stackdriver.

However I forgot to check for `gcp.distributed-sessions.enableTrace` in this class, I will correct that.

File: tomcat-gcp-lib/src/main/java/com/google/cloud/runtimes/tomcat/session/DatastoreSession.java [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+   * @param attributeKeyFactory The KeyFactory to use to create the key for the entity.
+   * @param name The name of the attribute to serialize.
+   * @return An Entity containing the serialized attribute.
+   */
+  private FullEntity serializeAttribute(KeyFactory attributeKeyFactory, String name) {
+    ByteArrayOutputStream bos = new ByteArrayOutputStream();
+    try (ObjectOutputStream oos = new ObjectOutputStream(bos)) {
+      oos.writeObject(getAttribute(name));
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
```
Comment (balopat) [inline tomcat-gcp-lib/src/main/java/com/google/cloud/runtimes/tomcat/session/DatastoreSession.java:171]: The exception handling is not covered by tests, please write tests for it.

File: tomcat/src/main/tomcat-base/conf/catalina.properties [hunk 7]
```diff
@@ -154,9 +154,13 @@ gae.proxy.trusted=".*"
...
+gcp.distributed-sessions.separateAttributes=false
 
 # Specify which Uri to ignore when persisting sessions.
 gcp.distributed-sessions.uriExcludePattern=
 
 # Set the attribute compression of the HTTP Connector
-tomcat.server.connector.compression=off
\ No newline at end of file
+tomcat.server.connector.compression=off
+
```
Comment (balopat) [inline tomcat/src/main/tomcat-base/conf/catalina.properties]: nit: remove new lines :)

File: tests/test-war-integration/src/main/java/com/google/cloud/runtimes/tomcat/test/integration/CustomServlet.java [hunk 8]
```diff
@@ -33,14 +35,11 @@
...
-    // Dump the server configuration into the test logs
-    configuration.append("{\"path\": \"dump/all\"}");
-    configuration.append("]");
+
+    String configuration = CharStreams.toString(
+        new InputStreamReader(getClass().getResourceAsStream("/custom-test-specification.json")));
 
     resp.setContentType("application/json");
-    resp.getWriter().print(configuration.toString());
+    resp.getWriter().print(configuration);
```
Comment (balopat) [inline tests/test-war-integration/src/main/java/com/google/cloud/runtimes/tomcat/test/integration/CustomServlet.java]: +1

**Review by balopat — CHANGES_REQUESTED** [review]

File: tomcat-gcp-lib/src/main/java/com/google/cloud/runtimes/tomcat/session/DatastoreSession.java (hunk 6 above)
Comment (cassand) [reply tomcat-gcp-lib/src/main/java/com/google/cloud/runtimes/tomcat/session/DatastoreSession.java:171]: Indeed, I added a test for serialization exception.

**Comment by cassand** [comment]
A subset of this pull request is available at #105, in which the possibility to use a single entity will be removed.

**Pull request closed without merging**